### ISSUE TITLE
[BUG] Inconsistent results of MLX exp function compared to other libraries

### ISSUE THREAD

**Issue by kemchenj** [ml-explore/mlx#1341]
**Describe the bug**

I have observed some subtle differences in the results of MLX's `exp` function compared to other libraries, and even between its CPU and GPU computations.

**To Reproduce**

Include code snippet

```python
import numpy as np
import torch as tf
import mlx.core as mx
import math

print(np.__version__)  # 2.0.0
print(tf.__version__)  # 2.3.1
print(mx.__version__)  # 0.16.1

v = float(-7.773141)
arr = np.array([v])

print(v)                         # -7.773141
print(arr.item())                # -7.773141
print(tf.from_numpy(arr).item()) # -7.773141
print(mx.array(arr).item())      # -7.773140907287598

print(math.exp(v))                                 # 0.0004208891755120154
print(np.exp(arr).item())                          # 0.0004208891755120154
print(tf.exp(tf.from_numpy(arr)).item())           # 0.0004208891755120154
print(mx.exp(mx.array(arr), stream=mx.cpu).item()) # 0.00042088920599780977
print(mx.exp(mx.array(arr), stream=mx.gpu).item()) # 0.0004208892351016402
```

**Expected behavior**

1. The `exp` function in MLX should produce results consistent with other widely-used libraries such as NumPy and PyTorch, at least to a high degree of precision (e.g., matching to 10-12 decimal places).
2. The results of MLX's `exp` function should be consistent across different compute devices (CPU and GPU) for the same input.
3. The conversion of floating-point numbers to MLX arrays should preserve the exact value without any loss of precision.

In general, we expect numerical operations in MLX to produce results that are either identical to or extremely close to those of other established scientific computing libraries, ensuring reproducibility and consistency across different platforms and libraries.

**Desktop (please complete the following information):**

 - OS Version: macOS 15.1 Beta (24B5024e)
 - Version: MLX 0.16.1

**Additional context**

While these differences are small, they could potentially lead to inconsistencies in certain applications, especially those requiring high precision or when accumulating errors over many operations.

**Comment by awni** [comment]
The NumPy, `math` and `tf` implementations are all in `float64` whereas MLX uses `float32`. If you compare the other packages in `float32` then you will get similar approximation errors as MLX in `float32`. 

Using `np.exp` on `float64` as the ground truth, here is what the approximation errors look like for `np.float32`, and `mx.float32` on the CPU and GPU.

You can see they are all pretty similarly accurate.

### Numpy fp64 vs fp32

![nprel](https://github.com/user-attachments/assets/16f1bc46-8e7b-45c6-83bd-0547be3f6616)

### Numpy fp64 vs MLX fp32 GPU

![mxrel_gpu](https://github.com/user-attachments/assets/f3dd71b9-f079-4128-8a11-2f30771857d0)

### Numpy fp64 vs MLX fp32 CPU
![mxrel_cpu](https://github.com/user-attachments/assets/1d7d3692-db35-44bd-b726-3f7b25bab35b)

Code for reference

```python
import mlx.core as mx
import numpy as np
import matplotlib.pyplot as plt

x = np.arange(-10, 10, 0.01)
gt = np.exp(x)

def relerr(pred):
    return np.abs((gt - np.array(pred)) / gt)

plt.plot(x, relerr(np.exp(x.astype(np.float32))))
plt.plot(x, relerr(mx.exp(mx.array(x))))
plt.plot(x, relerr(mx.exp(mx.array(x), stream=mx.cpu)))
```
